js/src/wrappers/oai.ts

```diff
@@ -348,83 +348,131 @@ function parseChatCompletionParams<P extends ChatParams>(
       metadata: spanInfoMetadata,
     },
   };
-  const { messages, ...paramsRest } = params;
-  return mergeDicts(ret, { event: { input: messages, metadata: paramsRest } });
+  const input = params[inputField];
+  const paramsRest = { ...params };
+  delete paramsRest[inputField];
+  return mergeDicts(ret, { event: { input, metadata: paramsRest } });
 }
 
-type EmbeddingCreateParams = {
-  input: string;
-};
-
-type CreateEmbeddingResponse = {
-  data: { embedding: Array<number> }[];
-  usage:
-    | {
-        total_tokens: number;
-        prompt_tokens: number;
-      }
-    | undefined;
-};
-
-function wrapEmbeddings<
-  P extends EmbeddingCreateParams,
-  C extends CreateEmbeddingResponse,
->(
-  create: (params: P, options?: unknown) => APIPromise<C>,
-): (params: P & SpanInfo, options?: unknown) => Promise<any> {
-  return async (allParams: P & SpanInfo, options?: unknown) => {
+function createApiWrapper<T, R>(
+  name: string,
+  create: (params: T, options?: unknown) => APIPromise<R>,
+  processResponse: (result: R, span: Span) => void,
+  parseParams: (params: T & SpanInfo) => StartSpanArgs,
+): (params: T & SpanInfo, options?: unknown) => Promise<any> {
+  return async (allParams: T & SpanInfo, options?: unknown) => {
     const { span_info: _, ...params } = allParams;
     return traced(
       async (span) => {
-        // We could get rid of this type coercion if we could somehow enforce
-        // that `P extends EmbeddingCreateParams` BUT does not have the property
-        // `span_info`.
         const { data: result, response } = await create(
-          params as P,
+          params as T,
```

Comment:
Can we get rid of this type coercion if we change the signature of `create` to this?

```
create: (params: Omit<T & SpanInfo, "span_info">, options?: unknown) => APIPromise<R>,
```

Reply:
done